### DISCUSSION TITLE
trying to writ a script to get a Keithley 2450 to out put a series of voltages and record the current

### DISCUSSION THREAD

**Discussion by zaidic** [pymeasure/pymeasure#438] [Q&A]
Hi All,

So what I'm trying to do is apply some range of voltages and measure the current back. My code :

```python

# import any useful extentions for controlling Keith
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import pymeasure
from time import sleep
from scipy.stats import linregress


# you write the code you can name em
from pymeasure.instruments.keithley import Keithley2450
keith = Keithley2450("USB0::0x05E6::0x2450::04455537::0::INSTR")

# resistance trace, +/- 0.5v get a Res value
def res_trace(n= None , max_v = None, min_v = None, data_points = None, r_crit = None):
    
    # governing parameters
    data_points = data_points or 100
    max_v = max_v or 0.5
    min_v = min_v or -max_v
    r_crit = r_crit or 1e+7
    av = 3
    
    # create arrays for storing data
    v = np.linspace( min_v, max_v, num = data_points)
    i = np.zeros_like(v)
    res_run = np.zeros_like(v) 
    
    # Keith!
    # set up for V app
    keith.reset()
    keith.use_front_terminals()
    keith.measure_current()
    keith.auto_range_source()
    keith.apply_voltage(voltage_range = (max_v - min_v))
    sleep(0.1)    
    keith.config_buffer( points = 3, 0)
    
    # actualy run
    for j in range (data_points):
        
        keith.ramp_to_voltage(v[j],1)
        keith.reset_buffer()
        sleep(15e-3)
        
        keith.start_buffer()
        keith.wait_for_buffer()
        
        i[j] = keith.mean_current
        
    ###
    
    # turn off 
    keith.shutdown()
    
    # add the data to the pandas file.
    
    
    # plot data
    plt.figure(1)
    plt.title(' Res Trace #'+n)
    plt.legend(loc='upper left', frameon=False)
    plt.xlabel('V')
    plt.ylabel('I')
    plt.plot(v,i)
    
    # calculate the res value and decide wether to continue
    r_stat, intercept, r_value, p_value, std_err = linregress(i, v)
    
    if r_stat > r_crit :
        r_quit = 1
    
    
    return r_quit, res_data
###

```

I'm having some trouble figuring out the process flow for what I want to do. In other word's the order I need to set up each part by. The py measure example isn't particularly helpful to me sink its the opposite process and trying to just flip them to get what I want doesn't appear to work.

When I run this code I get two errors. 1 is that there's no header names, the other is that there is no trigger function. I also get a pop up from the Keithley asking me to use the back as the current is so small its not accurate.
I also noticed if I put it in a :
```python

n=0
while  n!= 1 :
   res_trace()
   n = n +1 
```
It just runs indefinitely. Whenever I stop the code it reports having gotten to wait buffer. During this entire time the keithley doesn't read any values out.

I'd be incredibly appreciative to any help/ info you may have for me.

Z

**Comment by bilderbuchi** [comment]
Welcome! I am not familiar with the handling of Keithley buffers, but can offer a couple of remarks: 
* thanks for the paraphrased errors, but it's very hard to diagnose this. It's more helpful if you show the precise messages and where in the code they occurred.
* your wait_buffer invocation has the default timeout of 60 s - did you wait that long? maybe try setting a shorter time so that wait_buffer terminates earlier. 
* Is the command used in is_buffer_full the same for your instrument? As I understand it, the buffer should definitely become full at the end of the measurement, correct?

I'll also move this into the discussions board, as it is more of a usage question.
